architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,381 @@
+- Build database layer for Radius based on Dapr statestore (off by default).
+- Update UCP integration tests to use Dapr statestore instead of etcd.
+- Build support for automatic component generation for UCP. 
+  - Acceptance criteria: UCP can complete its startup sequence in the debugger with no Kubernetes, only Dapr.
+
+*These work items is being frontloaded to increase productivity of the remaining work. When this is done will have good workflows for development and for integration tests.*
+
+Stage 1:
+
+- Add Dapr installation to `rad init` and `rad install`.
+- Add Dapr annotations to our Kubernetes components.
+- Add default components to our Helm chart.
+- Add startup-time verification to our services to verify correct configuration of Dapr and its components.
+
+Stage 2: 
+
+- Switch our Go-based services to use the Dapr state store.
+- Remove etcd-based, crd-based, and cosmosdb-based database implementations.
+
+Stage 3:
+
+- Implement new async-worker.
+  - Acceptance criteria: works in UCP integration tests with no Kubernetes, only Dapr.
+- Switch our Go-based services to use the new async-worker.
+- Remove queue abstraction and implementations from Radius.
+
+Stage 4:
+
+- Add Dapr sidecar management and startup-time verification to DE.
+- Replace DE in-memory data layer with Dapr state store, update integration tests.
+
+## Open Questions / FAQ
+
+### Decisions
+
+*These are topics in-scope of this document, where we've had enough discussion to make a decision.*
+
+**Q: Can we use the Dapr state store as a database? What are the tradeoffs?**
+
+The Dapr state store building block provides the capabilities of a generic key-value-store. This is lower level than a document store like MongoDB or a RDBMS like PostgreSQL. 
+
+This isn't bad, we designed Radius around a key-value-store abstraction. Our existing database abstraction is similar to Dapr's state store. It's a good fit.
+
+There's one minor issue to address which is how to implement "list" operations (eg: list all containers in resource group X). Dapr state store doesn't have list functionality (though most key-value-stores do). The well-known solution for this is to maintain our own "index" as it's own key. This is a common enough solution that we shouldn't be concerned. 
+
+----
+
+*Follow-up: Why don't we use the Query building block in Dapr?*
+
+The query building-block is [deprecated](https://github.com/dapr/dapr/issues/7338#issuecomment-2132282783), and supported by a limited set of state stores. The query building-block will not see more investment, we should not take a dependency.
+
+----
+
+*Follow-up: What do we miss out on since we're not using an RDBMS?*
+
+RDBMS'es are useful when a system needs rich and flexible query capabilities, and the ability to enforce referential integrity across entities. These aren't requirements that we have in Radius. It's very hard to predict the future, and this design-document does not try.
```

Comment:
Did we say that as long as we use Dapr and provide the proxy layer, we can choose anything we want that the Dapr will use behind the scenes? Whether it is a key-value store or RDBMS.

Reply:
Yes.